forecasting/src/autogluon/forecasting/models/presets.py

```diff
@@ -0,0 +1,209 @@
+                ),
+            },
+        },
+    }
+    return default_model_hps[key]
+
+
+def get_preset_models(
+    freq: str,
+    prediction_length: int,
+    path: str,
+    eval_metric: str,
+    hyperparameters: Union[str, Dict],
+    hyperparameter_tune: bool,
+    invalid_model_names: List[str],
+    **kwargs,
+):
+    """
+    Create a list of models according to hyperparameters. If hyperparamaters=None,
+    will create models according to presets.
+    """
+    models = []
+    if isinstance(hyperparameters, str):
+        hyperparameters = copy.deepcopy(
+            get_default_hps(hyperparameters, prediction_length)
+        )
+    else:
+        hp_str = "default" if not hyperparameter_tune else "default_hpo"
+        default_hps = copy.deepcopy(get_default_hps(hp_str, prediction_length))
+
+        if hyperparameters is not None:
+            # filter only default_hps for models with hyperparameters provided
+            default_hps = {
+                model: default_hps.get(model, {}) for model in hyperparameters
+            }
+            for model in hyperparameters:
+                default_hps[model].update(hyperparameters[model])
+        hyperparameters = copy.deepcopy(default_hps)
+
+    if hyperparameter_tune:
+        verify_contains_searchspace(hyperparameters)
+    else:
+        verify_no_searchspace(hyperparameters)
+
+    invalid_model_names_set = set(invalid_model_names)
+
+    for model, model_hps in hyperparameters.items():
+        if isinstance(model, str):
+            if model not in MODEL_TYPES:
+                raise ValueError(f"Model {model} is not supported yet.")
+            model_type = MODEL_TYPES[model]
+        elif isinstance(model, AbstractForecastingModelFactory):
+            model_type = model
+        elif not issubclass(model, AbstractForecastingModel):
+            logger.warning(
+                f"Customized model {model} does not inherit from {AbstractForecastingModel}"
+            )
+            model_type = model
+        else:
+            logger.log(20, f"Custom Model Type Detected: {model}")
+            model_type = model
+
+        model_type_kwargs = dict(
+            path=path,
+            freq=freq,
+            prediction_length=prediction_length,
+            eval_metric=eval_metric,
+            hyperparameters=model_hps,
+            **kwargs,
+        )
+
+        # add models while preventing name collisions
+        model = model_type(**model_type_kwargs)
+        num_increment = 2
+        model_type_kwargs.pop("name", None)
+        while (
+            model.name in [m.name for m in models]
+            or model.name in invalid_model_names_set
+        ):
+            model = model_type(
+                name=f"{model.name}_{num_increment}", **model_type_kwargs
+            )
+            # increment only if model name collided in this round
+            if model.name not in invalid_model_names_set:
+                num_increment += 1
```

Comment:
Don't you mean `in`? Won't this infinite loop otherwise? Plus, this won't work if the collision is in `models` and not in `invalid_model_names_set`.

Further, simpler to just always `num_increment += 1` at start of while loop and have `num_increment = 1` initially, avoids having to check twice.

Reply:
The loop itself doesn't depend on `num_increment` and if there is a collision it just keeps appending `_{num_increment}` so I don't think it would infinite loop in any case. The logic, and why it doesn't add to `invalid_model_names_set` as suggested above is as follows.

`num_increment` is only incremented if the name collision happened this round. So if 
```
invalid_model_names (those from previous calls to fit): ["A", "B"]
new default model names being pushed: ["B", "B"]
the output is ["B_2", "B_3"]
```
and 
```
invalid_model_names (those from previous calls to fit): ["A", "B", "B_2"]
new default model names being pushed: ["B", "B"]
the output is ["B_2_2", "B_2_3"]
```
If we didn't perform the check on L175, then the outputs would be `"B_2_3", "B_2_4"`.
Logic looks OK to me but yes I think it will pay to make it more legible.

Reply:
Thanks for looking into this. It is nuanced but actually pretty important.

Please refer to my new comment discussing what I think can be improved from the current implementation.

In your example, my suggested implementation instead of:

```
invalid_model_names (those from previous calls to fit): ["A", "B", "B_2"]
new default model names being pushed: ["B", "B"]
the output is ["B_2_2", "B_2_3"]
```

would be:

```
invalid_model_names (those from previous calls to fit): ["A", "B", "B_2"]
new default model names being pushed: ["B", "B"]
the output is ["B_3", "B_4"]
```